Close #6525: linkcheck: Add linkcheck_ignore_redirects and linkcheck_warn_redirects (#9234)
sphinx-doc/sphinx: doc/usage/configuration.rst

   .. versionadded:: 1.1

> .. confval:: linkcheck_ignore_redirects

Reviewer: Rename this to `linkcheck_allowed_redirects`?

"Ignored" may imply that it's not followed and I think it should be checking the target of the redirect too.

Reviewer 2: Yes. Sorry, that was my bad wording. I'm not sure that "allowed" is any better as that implies that other redirects are disallowed.

I'd suggest `linkcheck_silenced_redirects`. That makes clear that we're just silencing `'redirected'` and treating it as `'working'`.

Author: I feel "silenced" surely means suppressing warnings. But, I feel it does not mean changing `'redirected'` to `'working'`. So I prefer to use "allowed" instead of "ignore".

   A dictionary that maps a pattern of the source URI to a pattern of the canonical
   URI.  If set, linkcheck builder treated as "ignored" if a hyperlink in the source
   document that matches to the pattern of the source URI, and it redirects to the

>    URL that matches to the pattern of the canonical URI.  Example:

Reviewer 3: ```suggestion
   A dictionary that maps a source URI pattern to a canonical URI pattern. The
   linkcheck builder treats the redirected link as "ignored" when:

    - the link in the document matches the source URI pattern, and
    - the redirect location matches the canonical URI pattern.

   Example:
```

Reviewer 2: Based on my other comments, we should change `"ignored"` → `"working"` and refer to "silencing" the redirect.

   .. code-block:: python

      linkcheck_ignore_redirects = {
          # All HTTP redirections from the source URI to the canonical URI will be ignored.

>           'http://sphinx-doc.org/.*': 'https://sphinx-doc.org/en/master/.*'

Reviewer: I don't think the example `'https://sphinx-doc.org/.*': 'https://sphinx-doc.org/en/master/.*'` is good because it allows any redirect without checking that the page is actually the same.

For the specific case of `https://sphinx-doc.org/`, a redirect to anything in `https://(www\.)?sphinx-doc.org/.*` would be ok because the intent is just to link to the root page of the website.

Reviewer: Since these are regexps it would have to be `r'https://sphinx-doc\.org/.*': r'https://sphinx-doc\.org/en/master/.*'` too...

Reviewer 2: I'd make a small tweak to anchor it and drop the `.*`:

```suggestion
          r'^https?://sphinx-doc\.org/': r'^https://sphinx-doc\.org/en/master/'
```

(Obviously this will need updating if we need to use the capturing group approach.)

      }

   It's helpful to enable :confval:`linkcheck_warn_redirects` not to warn the URIs

>    that causes HTTP redirection as expected.

Reviewer 3: ```suggestion
   It's helpful to enable :confval:`linkcheck_warn_redirects` to warn for URIs
   causing unexpected HTTP redirection.
```

   .. versionadded:: 1.1

.. confval:: linkcheck_allowed_redirects

   A dictionary that maps a pattern of the source URI to a pattern of the canonical
   URI. The linkcheck builder treats the redirected link as "working" when:

    - the link in the document matches the source URI pattern, and
    - the redirect location matches the canonical URI pattern.

   Example:

   .. code-block:: python

>       linkcheck_working_redirects = {

Reviewer 3: ```suggestion
      linkcheck_allowed_redirects = {
```

          # All HTTP redirections from the source URI to the canonical URI will be treated as "working".

>           r'https://sphinx-doc\.org/.*': r'https://sphinx-doc\.org/en/master/.*'

Reviewer: If there's no regex capture then this isn't checking what it implies it's checking. To allow for a more complete example without comparing the source and destination paths, I've added an allowance for redirecting to `www.` but this could be removed too.

```suggestion
          r'https://sphinx-doc\.org/': r'https://(www\.)?sphinx-doc\.org/.*'
```

Author: I think the example should use patterns for both source and canonical URIs. So I think the current one is enough to describe the feature.

Reviewer: But this is not just allowing redirects to the canonical version of a URL, it's allowing any redirects.

You may as well use `r'https://sphinx-doc\.org/.*': r'https://sphinx-doc\.org/.*'` because there isn't really any validation of the destination.

Author: What do you mean? The current rule verifies the pages under `https://sphinx-doc.org/` to the URL under `https://sphinx-doc.org/en/master/`.  It does not mean "any redirects".

Reviewer: It's allowing practically any redirect, so not what I would consider a good example.

It's not far off allowing `https://www\.example\.com/.*` to `https://www\.example\.com/.*` which then causes linkcheck to ignore removed pages that just redirect to the home page.

Author: If `https://sphinx-doc.org/foo/bar` redirects to `https://sphinx-doc.org/other/path`, it's not "allowed". Is this allowing any redirect? I'm confused.

   .. versionadded:: 1.1

.. confval:: linkcheck_allowed_redirects

   A dictionary that maps a pattern of the source URI to a pattern of the canonical
   URI. The linkcheck builder treats the redirected link as "working" when:

>     - the link in the document matches the source URI pattern, and
>     - the redirect location matches the canonical URI pattern.

Author: Bad mark-up:

```suggestion
   - the link in the document matches the source URI pattern, and
   - the redirect location matches the canonical URI pattern.
```

   .. versionadded:: 3.4

.. confval:: linkcheck_warn_redirects

   If true, emit a warning when redirection detected on checking hyperlinks.
   It's useful to detect unexpected redirects under :option:`the warn-is-error

>    mode <sphinx-build -W>`.  Default is ``False``.

Reviewer 3: ```suggestion
   If true, emits a warning when the response for a link is a redirect. It's
   useful to detect unexpected redirects under :option:`the warn-is-error mode
   <sphinx-build -W>`.  Default is ``False``.
```

   .. versionadded:: 3.4

.. confval:: linkcheck_warn_redirects

   If true, emits a warning when the response for a hyperlink is a redirect.
   It's useful to detect unexpected redirects under :option:`the warn-is-error
   mode <sphinx-build -W>`.  Default is ``False``.

>     .. versionadded:: 4.1

Author: Bad mark-up too:

```suggestion
   .. versionadded:: 4.1
```
